types/src/proof/mod.rs

```diff
@@ -486,6 +486,115 @@ pub fn verify_sparse_merkle_element(
     Ok(())
 }
 
+/// Computes the root hash of an accumulator using the root hashes of the frozen subtrees and their
+/// corresponding sibling nodes. As an example, given the following accumulator that has 10
+/// elements, the frozen subtree roots and the corresponding siblings are annotated below.
+///
+/// ```text
+///                                        root
+///                                      /      \
+///                                    /          \
+///                                  /              \
+///                                /                  \
+///                              /                      \
+///                            /                          \
+///                          /                              \
+///                        /                                  \
+/// frozen_subtree_roots[0]                                    o
+///                    /   \                                  / \
+///                   /     \                                /   \
+///                  o       o                              o     siblings[0]
+///                 / \     / \                            / \
+///                o   o   o   o    frozen_subtree_roots[1]   sibling[1]
+///               / \ / \ / \ / \                       / \
+///               o o o o o o o o                       o o
+/// ```
+///
+/// This function has at least two possible use cases:
+///   1) Given an accumulator represented by all of its frozen subtrees, we want to compute the
+///      root hash by combining the subtrees with placeholder nodes. In this case the caller can
+///      use a list of placeholder nodes as the `siblings` parameter.
+///   2) To prove that an accumulator A can be obtained by appending some elements to another
+///      accumulator B, the proof can provide a list of siblings and we can combine B's frozen
+///      subtrees with these siblings.
+pub fn get_accumulator_root_hash_by_frozen_subtrees_and_siblings<H: CryptoHasher>(
+    frozen_subtree_roots: &[HashValue],
+    num_leaves: u64,
+    siblings: &[HashValue],
+) -> HashValue {
+    // Special cases when there is zero or one subtree.
+    match frozen_subtree_roots.len() {
+        0 => return *ACCUMULATOR_PLACEHOLDER_HASH,
+        1 => return frozen_subtree_roots[0],
+        _ => (),
+    }
+
+    // For accumulators that have more than one frozen subtrees, the rightmost/smallest frozen
+    // subtree will always be combined with a sibling on the right.
```

Comment:
This part is too tricky. My thought is move line544 up with `+1` removed and merge line 536-540 into the while loop starting at line 549

Reply:
This part is special, so I think it makes sense to do it separately. Not sure how to merge it into the loop.

Reply:
if you do 
```
let mut bitmap = num_leaves >> (num_leaves.trailing_zeros()
```
Then the last bit of `bitmap` must be 1, which means line 550 is true and the first iteration of the loop is going to the be combining with a sibling always, right?

Reply:
@lightmark if last bit of `bitmap` is 1, it means line 550 is false and it would go to the other branch.